src/ApplicationModule.cc

```diff
@@ -99,13 +100,19 @@ namespace ChimeraTK {
   void ApplicationModule::mainLoopWrapper() {
     Application::registerThread("AM_" + className()); // + getName());
 
+    // normal ApplicationModules use exclusive testable mode lock for the startup phase for performance reasons, only
+    // DeviceManagers have to use a shared lock from the beginning to avoid deadlocks in the barriers.
```

Comment:
```suggestion
    // Testable mode: Determine whether a shared or an exclusive testable mode lock shall be used during the initial
    // startup phase of this ApplicationModule. Normal ApplicationModules use exclusive testable mode lock for the
    // startup phase for performance reasons, only DeviceManagers have to use a shared lock from the beginning to
    // avoid deadlocks in the barriers.
    // Once the actual mainLoop() is entered, we always use a shared lock, since it should be faster to allow parallel
    // execution of the ApplicationModules.
```

Reply:
I was more referring to the DeviceManager and deadlock in barriers part.

" DeviceManagers have to use shared locks because they will wait for other DeviceManagers in a barrier in their main loop while holding the testable mode lock. Once one DeviceManager has reached this point, no other DeviceManager is able to get an exclusive lock in its initialisation."

Reply:
I am sorry, I think that's too much explanation of general things, and your explanation actually is a bit short ("Once one DeviceManager has reached this point" - no, already before, that is the very nature of an exclusive lock). This might be actually misleading and confusing if the code is later reviewed under different aspects. IMO pointing to the deadlock potential w.r.t. the barriers is sufficient. During a later revisit one needs to think about the consequences to the problem at hand anyway. These comments are not there to justify the current change / commit but to allow for a simpler understanding of the code in a potentially distant future from now.

Reply:
I agree that the second sentence is too general.

>  "Once one DeviceManager has reached this point" no, already before, that is the very nature of an exclusive lock

It is only important at "this point" because that's where it is waiting in a barrier for another DeviceManager, which does not work with shared locks.

What bothers me most is "avoid deadlocks in the barriers", and "the barriers" is not known at this context. How about "their barriers"?

Reply:
Ok, I changed it into "their RecoveryGroup barriers".

Reply:
> It is only important at "this point" because that's where it is waiting in a barrier for another DeviceManager, which does not work with shared locks.

This is true only for our current problem at hand. In a future perspective, this simplification might be highly confusing...